### ISSUE TITLE
forAll with ||, and exists with &&, can be inconsistent

### ISSUE THREAD

**Issue by pford19** [rickynils/scalacheck#425]
**Summary**

Indivudual existential (`exists`) properties that succeed can fail when combined in a conjunction with `&&`.

Likewise individual universal (`forAll`) properties that fail can succeed when combined in a disjunction with `||`.

This is not the conventional expected semantics of `&&` and `||`, and probably not the intended semantics either.

The inconsistency can be demonstrated unambiguously (100% guaranteed) in a small specialized failing
case with just two properties and a fixed-order generator. See Example 1 below.

The inconsistency can be demonstrated in a slightly more realistoc scenario as well, involving a
larger number of properties based on random generators. With the random generators, the
inconsistency is highly probable, but not 100% guaranteed. See Example 2 below.

**Comments**

Here's my concpetual model of what is going on.

Assume 
- well-behaved generators that can be viewed as value streams
- a`Prop` is parameterized by a generator `g` and a predicate `f`.

Then a simplified conceptual model of a `Prop` is

	class Prop(g:Stream[T], f: T=>Boolean){
		def success: Boolean
	}
	def &&(p1: Prop, p2: Prop): Prop
	def ||(p1: Prop, p2: Prop): Prop
	
	
For an existential prop `success` is defined as

	def success = g.take(BIG_ENOUGH_SAMPLE).exists(f(_)) // existential case
	
For a universal prop `success` is defined as 

	def success = g.take(BIG_ENOUGH_SAMPLE).forall(f(_)) // universal case
	
The conceptual definition of success for the conjunction  `&&` of properties is

	def (p1 && p2).success = p1.success && p2.success  // (1)
	
But it appears that the implementation of `&&` for  _existential_ properties is more along these lines

	// p1 and p2 are both existential
	def (p1 && p2).success = (p1.g).zip(p2.g).map((v1, v2) => p1.f(v1) && p2.f(v2))).exists(_) // (2)

and the implementation of `||` for  _universal_ properties is more along these lines

	// p1 and p2 are both universal
	def (p1 && p2).success = (p1.g).zip(p2.g).map((v1, v2) => p1.f(v1) || p2.f(v2))).forall(_) // (3)

This model for `||` and `&&` is mostly reverse engineered from observed behavior and a bit of poking
around in the code. But the code is kind of hard to follow for the uninitiated. Experts should
confirm or enlighten.

If these conceptual models of conjunction of existential properties and disjunction of universal
properties are more or less correct then it easy to contrive examples for which (2) and (3) are
inconsistent with (1).

The key issue with (2) and (3) is that they evaluate the the combination of the predicates __per
iteration of the generators__, and then combine those single reduced values using `exists` or
`forall`. 

Consider two universal predicates `p1` and `p2` and assume the following Boolean streams of `g.map(f)`

	iteration : 1 2 3 4 ...
	p1        : F T T T ...
	p2        : T F T T ...
	p1 || p2  : T T T T ...
	
The `p1` stream and the `p2` stream each contain in `F`, so neither universal property succeeds. But
the per iteration disjunction of the two streams is all `T`, so `p1 || p2` succeeds, per model (3)
above.

Likewise, for existential properties `p1` and `p2` a complementatory scenario can be laid out.

	iteration : 1 2 3 4
	p1        : F T F F ...
	p2        : T F F F ...
	p1 && p2  : F F F F ...
	
Each of `p1` and `p2` succeeds as existential properties because the mapped predicate streams each
include a `T`. But the conjunction of the mapped streams is all `F`, so `p1 && p2` fails, per model
(2).

Based on the above schematic layouts, what could be called the "predicate matrix", the
implementations of `||` and `&&` can be described as column-first (or, equivalently,
iteration-first). 

As these examples illustrate, colum-first evaluation of the predicate streams can result in an
incorrect outcome. A correct implementation would be row-first evaluation of the predicate streams.

**Note on Current Implementation**

The current implementations of `&&` and `||` are correct in specific cases:

- Conjunction of universal properties (`&&` of `forAll`) gives the expected outcomes. This seems
  likely to be the most common case in practice.

- Disjunction of existential properties (`||` of `exists`) gives the expected outcomes.

In the former case (`&&` of `forAll`), the semantics of success require the entire predicate matrix
to be all `T`s. In that case row-first and column-first evaluations give the same outcome.

Likewise for `||` of `exists`, the semantics of success require the entire predicate matrix to have
at least one `T` in some position. In the case as well, row-first and column-first evaluation yield
give the same outcome.

**Workarounds**

For conjuction of existential properties a workaround is to create and test separate properties and
avoid the conjunction.

For disjunction of universal properties a workaround is to combine the predicates into a single
predicate wrapped in a single property. Of course, this requires a common generator.

**Detailed Example 1**

This example demonstrates the inconsistency 100% repeatably because it relies on fixed-order
non-random generators and props based on them which are carfully orchestrated to align (or misalign)
the predicate streams to demonstrate the inconsistency.

Example 2 

**Example 1 Code**
```
  // Factory for fixed order (true, false, ...) generators. 
  def makeTF(): Gen[Boolean] = {
    val c = List(true, false)
    var it = c.iterator
    Gen.const(-99).map { _ => if (!it.hasNext) it = c.iterator; it.next }
  }

  // Prop factory: asserts that some generated value == `b`. (Expected to succeed.)
  def makeExists(b: Boolean): Prop = exists(makeTF())(_ == b)

  // Prop factory: asserts that all generated values == `b`. (Expected to fail.)
  def makeForAll(b: Boolean): Prop = forAll(makeTF())(_ == b)

  // Each makeExists property succeeds individually.
  List(true, false).foreach { i =>
    val ep = makeExists(i)
    property(s"ep($i) succeeds individually") = ep
  }

  // Conjunction of successful existential properties fails.
  val epT = makeExists(true)
  val epF = makeExists(false)
  val conjunction = epT && epF
  property(s"ep(true) && ep(false) fails") = conjunction

  // Each makeForAll property fail individually.
  List(true, false).foreach { i =>
    val ap = makeForAll(i)
    property(s"ap($i) fails individually") = ap
  }

  // Disjunction of failing universal properties succeeds.
  val apT = makeExists(true)
  val apF = makeExists(false)
  val disjunction = epT || epF
  property(s"ap(true) || ap(false) succeeds") = disjunction

```
Notes on the code
- The `makeTF` factory method produces fixed order generators, that always generate the stream (true, false, true, false, ...).
- The `makeExists` and `makeForAll` factory product properties based on `makeTF` generators.
- Each property based on fixed order generators require fresh `Prop` and fresh `Gen` instances.
- The constructed `Props` are assigned to val before being set assigned to a property: See [https://github.com/rickynils/scalacheck/issues/424 Issue 424]

**Example 1 Output**

Here is the output (lightly edited) from checking the above properties.

```
+ ep(true) succeeds individually: OK, proved property.
+ .ep(false) succeeds individually: OK, proved property.
! ep(true) && ep(false) fails: Gave up after only 0 passed tests. 501 tests were discarded.

! ap(true) fails individually: Falsified after 1 passed tests.
! ap(false) fails individually: Falsified after 0 passed tests.
+ ap(true) || ap(false) succeeds: OK, proved property.
```

**Detailed Example 2**

This example demonstrates the inconsistency with very high probability but not 100%.

It is a bit more realisitic than the contrived Example 1. 

The combined `all` prop uses 10 props that separately check the the `choose(0,9)` does in fact, eventually, 
choose each integer in the range `(0 to 9)`. 

However for the `all` conjunction to succeed it requires that some specific iteration of the 10
generators for the 10 properties generate the precise sequence of values `(0, 1, 2, 3, 4, 5, 6, 7,
8, 9)` which has an extremely low probability.

**Example 2 Code**

```
  def makeChooseProp(n: Int) = exists(Gen.choose(0,9))(_ == n)

  (0 to 9).foreach { i =>
    val p = makeChooseProp(i)
    property(s"makeChooseProp($i) succeeds") = p
  }

  val props = (0 to 9).map(makeChooseProp)
  property("(0 to 9).map(makeChooseProp) generally fails") = all(props:_*)
```

**Example 2 Output**

```
+ makeChooseProp(0) succeeds: OK, proved property.
+ makeChooseProp(1) succeeds: OK, proved property.
+ makeChooseProp(2) succeeds: OK, proved property.
+ makeChooseProp(3) succeeds: OK, proved property.
+ makeChooseProp(4) succeeds: OK, proved property.
+ makeChooseProp(5) succeeds: OK, proved property.
+ makeChooseProp(6) succeeds: OK, proved property.
+ makeChooseProp(7) succeeds: OK, proved property.
+ makeChooseProp(8) succeeds: OK, proved property.
+ makeChooseProp(9) succeeds: OK, proved property.
! (0 to 9).map(makeChooseProp) generally fails: Gave up after only 0 passed tests. 501 tests were discarded.
```

**Comment by ashawley** [comment]
I tried re-writing the example, so that it had the property tests fully written out and avoid being overly DRY.  It's not clear if that's the reason for the inconsistency or not.  This also avoids assigning any `Prop`  to class variables, and just use the `Properties.update` method typical of most ScalaCheck suites.

```scala
  // Deterministic (true, false, ...) generator.
  val makeTF: Gen[Boolean] = {
    val it = Iterator.iterate(true)(!_)
    Gen.const(-99).map(_ => it.next)
  }

  // Each exists property succeeds individually.
  property("exists(_ == true) succeeds individually") = {
    Prop.exists(makeTF) {
      _ == true
    }
  }
  property("exists(_ == false) succeeds individually") = {
    Prop.exists(makeTF) {
      _ == false
    }
  }

  // Conjunction of successful existential properties fails.
  property("exists(_ == true) && exists(_ == false) succeeds") = {
    Prop.exists(makeTF) {
      _ == true
    } && Prop.exists(makeTF) {
      _ == false
    }
  }

  // Each forAll property fail individually.
  property("forAll(_ == true) fails individually") = Prop.forAll(makeTF){
    _ == true
  }
  property("forAll(_ == false) fails individually") = Prop.forAll(makeTF){
    _ == true
  }

  // Disjunction of failing universal properties succeeds.
  property("forAll(_ == true) || forAll(_ == false) fails") = {
    Prop.forAll(makeTF){
      _ == true
    } && Prop.forAll(makeTF){
      _ == false
    }
  }
```

**Comment by ashawley** [comment]
Looks like a mix-up in the example, but I don't think it changes your conclusions:

```patch
@@ -30,7 +30,7 @@
   }
 
   // Disjunction of failing universal properties succeeds.
-  val apT = makeExists(true)
-  val apF = makeExists(false)
-  val disjunction = epT || epF
+  val apT = makeForAll(true)
+  val apF = makeForAll(false)
+  val disjunction = apT || apF
   property(s"ap(true) || ap(false) succeeds") = disjunction

```

**Comment by pford19** [comment]
I'll have to remember `Iterator.iterator`. There is a lot of advanced and elegant vocabulary in the collections libraries to learn.

You are right, the code sample in my post is wrong. I was trying to be very careful to paste in  executed code, but I guess I got careless. I'll update the original post.

You also point out that you use `Property.property.update` directly rather than first assigning each `Prop` expression to a `val`. I did that on purpose because of odd behavior I was seeing with fixed-order deterministic generator passed as function value to `update` -- namely that the property check constructed a new generator instance for each tested value. See separate issue [https://github.com/rickynils/scalacheck/issues/424 Issue 424]. 

It seems like your variation uses a single shared generator instance, encapsulating a single shared iterator, for all the Props.  Have to think on that, but it seem like that coupling further constrains the interaction of generator streams for the the `&&`'ed or `||`'ed props.  For the combined props it is possible that the two component `Props`, `p1` and `p2`, "take turns" advancing the shared generator, resulting in generated predicate streams that look something like this.
```
 p1      : T T T T ...
 p2      : F F F F ...
 p1 || p2: T T T T ...
 p1 && p2: F F F F ...
```

**Comment by pford19** [comment]
OK. I've skinnied the failing example down to a minimal case involving `exists && exists`.

It relies on fixed-order, non-random, generators, and the careful management of vals and instances.

I'm reasonably confident that issues with `forAll || forAll` are complementary and both issues are
probably rooted in how `Prop.combine` works.

Analysing this one case fully, first, probably leads to a fuller understanding of how all the
logical property combinators _do_ work and _should_ work.

Here's the code.

```
/**
  * Minimal demonstration that Prop && of two successful existential props
  * can fail.  Depends on careful synchronization of fixed-order (non-random generators.)
  */
object ConjunctionCounterExample extends Properties(">>>>>>>>>>>>>>>>>") {

  override def overrideParameters(p: Parameters) =
    p.withMinSuccessfulTests(4)
      .withMaxDiscardRatio(1F)

  def loggedGen[T](tag: String, g: Gen[T]): Gen[T] = {
    var j = 0;
    g.map { v => j += 1; println(s"$tag($j)=$v"); v }
  }
  
  // Infinite (true, false, ...) iterator.
  def makeTF(): Gen[Boolean] = {
    var it = Iterator.iterate(true)(!_)
    Gen.const(-99).map { _ => it.next }
  }

  // Use distinct generator vals (that generate equivalent (T, F, ...) streams).
  //
  // Using vals ensures that the same generator instance is reused for each test of the two
  // component exists props.  Otherwise the delayed lazy Prop encapsulated by Properties.update may
  // create a new generator instance for each test.
  // 
  // Using two generators ensures that tests of the individual component props advance the
  // generators independently.

  val genP1 = loggedGen("p1", makeTF)
  val genP2 = loggedGen("p2", makeTF)

  // Prop p1 is true for odd iterations of genP1 (1, 3, 5, etc.), false for even iterations (2, 4,
  // 6, ...).  Conversely, Prop p2 is false for odd iterations of genP2, true for even iterations.
  // Thus for any given iteration `i` of the generators, `p1(i) && p2(i) == false`.

  val p1 = exists(genP1)(_ == true)
  val p2 = exists(genP2)(_ == false)

  // The following && property fails even though p1 and p2 independently succeed.
  // 
  // It "appears" that the results of each generator iteration are first combined with &&, which
  // results in a constant stream of `false`.  It then "appears" that the false stream is treated as
  // results for an exists Prop and fails.
  // 
  // I use "appears" in quotes, because the observed behavior is consistent with that description,
  // but I have not yet successfully traced the execution logic from the source code.


  property("x") = p1 && p2
}

```



The following is the scalacheck console output of the above code, showing the detailed logging from the
two generators. `p1(j)` is the `j`th iteration of the `Prop p1` generator. Likewise `p2(j)` is the
`j`th iteration of the `p2` generator.

```
p1(1)=true
p2(1)=true
p1(2)=false
p2(2)=false
p1(3)=true
p2(3)=true
p1(4)=false
p2(4)=false
p1(5)=true
p2(5)=true

>>>>>>>>>>>>>>>>> Gave up after only 0 passed tests. 5 tests were discarded.
Found 1 failing properties.
```

The generator log can be converted into the following table. `gx` denotes the generated values for
`Prop p1` or `Prop p2`. Likewise `fx` denotes the `Prop` predicate applied to the generated value. The `f1&&f2` row is the per-iteration logical `&&` of the `f` values.


```
          generator
          iteration
    
          1 2 3 4 5 
        +----------
 p1  g1 | t f t f t
     f1 | t f t f t  ==> p1 succeeds because the f1 stream includes a t
        |----------
 p2  g2 | t f t f t
     f2 | f t f t f  ==> p2 succeeds because the f2 stream includes a t
        |----------
 f1&&f2 | f f f f f  ==> apparent source of failure of p1&&p2
```

**Comment by ashawley** [comment]
> It seems like your variation uses a single shared generator instance, encapsulating a single shared iterator, for all the Props. Have to think on that, but it seem like that coupling further constrains the interaction of generator streams

You're right, I should have kept it a `def`, so that each generator had its own iterator, rather than sharing the iterator.

> I've skinnied the failing example down to a minimal case involving

I've tried to change your example again to use the `Property.property.update`, under the naive assumption that it may have some bearing on your bug because of some interaction between initialized class members and delayed evaluation of properties.  However, that seems to be entirely superstitious on my part.  Here's my rewrite of your "skinnied" example:

```scala
  def loggedGen[T](tag: String, g: Gen[T]): Gen[T] = {
    var j = 0;
    g.map { v => j += 1; println(s"$tag($j)=$v"); v }
  }
  
  def makeTF(): Gen[Boolean] = {
    var it = Iterator.iterate(true)(!_)
    Gen.const(-99).map { _ => it.next }
  }

  val genP1 = loggedGen("p1", makeTF)
  val genP2 = loggedGen("p2", makeTF)

  property("x") = {
    Prop.exists(genP1) {
      _ == true
    } && Prop.exists(genP2) {
      _ == false
    }
  }
```

I got the same output:

```
> testOnly ConjunctionCounterExample -- -s 4 -r 1.0
p1(1)=true
p2(1)=true
p1(2)=false
p2(2)=false
p1(3)=true
p2(3)=true
p1(4)=false
p2(4)=false
p1(5)=true
p2(5)=true
[info] ! >>>>>>>>>>>>>>>>>.x: Gave up after only 0 passed tests. 5 tests were discarded.
```

**Comment by ashawley** [comment]
> The main takeaway, detailed below, is that both `exists` props are repeatedly evaluated, even after each has been proved (`Prop.Result.status == Proved`).

The algebra for ScalaCheck properties doesn't work like traditional Boolean logic in programming.  For Boolean expressions, the outcome of the first condition could be short-circuited for subsequent conditions or for certain Boolean operators they are eagerly evaluated.  It seems you wish this was the case for ScalaCheck properties.

> It "appears" that the results of each generator iteration are first combined with &&, which results in a constant stream of `false`.

Indeed, it seems the current implementation in ScalaCheck will combine two properties with `&&` in to one, and then the generators are run for each of the separate properties simultaneously to produce a pair of values where both properties are true.   I understand the confusion from properties becoming combined and then tested, rather than tested atomically and then, and only then, their results combined.  This could be especially confusing if you assign your properties to variables,  as you have done, which are later be combined in an actual property.  You've been writing independent properties that are combined.  Maybe there would be a way for ScalaCheck to smartly run properties separately when they are combined together like this.  However, not all properties are independent.  There are ways in ScalaCheck to compose properties such that they share scope and need to be evaluated together.

**Comment by pford19** [comment]
@ashawley: Thanks for your comments. I edited the original post about 10 days based on further analysis of the code with, what I believe, is a more thorough analysis of the source of the surprising behavior.

> The algebra for ScalaCheck properties doesn't work like traditional Boolean logic in programming.

I understand that a Prop is not a simple binary true/false. With a single test it may be proved, disproved, undecided (or throw an exception). After sufficient multiple tests it may by proved, plausible, implausible, disproved, or exhausted. Proved and plausible act more or less like true ("true-ish"), and disproved and implausible and exhausted act more or less like false ("false-ish").

> the outcome of the first condition could be short-circuited for subsequent conditions or for certain Boolean operators they are eagerly evaluated. It seems you wish this was the case for ScalaCheck properties.

What I wish is that the actual results of `&&` and `||` make sense: that is, that two "true-ish" properties, when combined with `&&` result in a "true-ish" property, and two "false-ish" properties, when combined with `||` result in a "false-ish" property.

 **How** that is done is an implementation detail, but it should be **logically equivalent** to evaluating the two properties sequentially. Having two "true-ish" properties combined with `&&` and result in a "false-ish" property seems **fundamentally** wrong.

>  I understand the confusion from properties becoming combined and then tested.

It's more than a matter of confusion about the semantics of combined properties.  This implementation approach results in behavior that is surprising, non-sensical, and arguably **wrong**.  (Of course, software is what software does, regardless of what it is named or what a doc-string implies. However, in this case the semantics of `&&` (what `&&` does) as currently implemented, seems tricky to articulate in general, not very useful for any case but `forAll && forAll`, and most probably not what was originally intended.)

What seems missing in the implementation of the test loop for a compound property is any mechanism to "remember" that a nested component property of top-level compound property has been proved or disproved (decided unambiguously).

Once a nested component property has been decided it should no longer be subject to additional tests (or additional tests should not affect the decision). As currently implemented, continuing to test such a property effectively "forgets" the unambiguous decision  and puts the component property back in an undecided state.

The test loop is stateful for the top-level compound property (it maintains full history of all the tests) and it handles early termination of the test loop for unambiguous decisions. 

But nested component properties are hidden from the test loop and do not benefit from the stateful logic in the top-level test loop. 

More stateful properties might be one way to address this issue, but that seems pretty complicated in the current implementation.  For instance, 
- `Prop.apply` returns a new `Prop` rather than just a `Result`, 
- a `Prop` remembers its complete test history
- in particular, a Prop remembers that it has been unambiguously decided no matter how many more times it is tested

> However, not all properties are independent. There are ways in ScalaCheck to compose properties such that they share scope and need to be evaluated together.
 
I'd be interested to see a concrete example of such a case.

**Comment by pford19** [comment]
@ashawley, thanks for the examples. 

The **first example** does not appear to an example of `Prop.||`. Isn't it just a plain old vanilla Boolean `||`, since `Prop.throws` returns `Boolean`?

A minor point is that the property would be more precisely expressed as 
```
 inside != Prop.throws(classOf[IndexOutOfBoundsException])(xs(i))
```
As written, the property passes if both `inside == true` and `xs(i)` throws an exception. That is, the strongest expression of the property is EXCLUSIVE-OR (`!=`), rather than EITHER-OR-BOTH (`||`).

The **second example** illustrates some of the rich ergonomics supported by `Props`. But ...

- Glossing over the details of labeling magic, these are constant props that wrap a boolean and always return proved or disproved results. 
- Such constant props are quite well behaved with respect to this issue: compounding them with other properties using `&&` or `||` will not contribute to surprising outcomes since they give the same result every time they are tested.

In your previous post you wrote: 
> However, not all properties are independent. There are ways in ScalaCheck to compose properties such that they share scope and need to be evaluated together.

The second example shows composed properties that share scope, but all constant.

The following examples, illustrating the issue, follows your lead:
- Props are not assigned to `vals` (not that I think that should make any difference to the semantics)
- Props are nested and depend on enclosing scope

```
  property("contrived that succeeds") = Prop.forAll(Gen.choose(1,2)) { n: Int =>
    Prop.forAll(Gen.choose(1,n)) { _ <= 2 } &&
      Prop.exists(Gen.choose(1, 2)) { _ == n }
  }
  property("contrived that fails") = Prop.forAll(Gen.choose(1,8)) { n: Int =>
    Prop.forAll(Gen.choose(1,n)) { _ <= 8 } &&
      Prop.exists(Gen.choose(1, 8)) { _ == n }
  }
```

I'm still pretty convinced that  
- the fundamental problem is sufficiently demonstrated without the need to consider more complex cases with nested properties and shared scope
- if there is any way to address the fundamental problem in the "simple" case then more complex cases (as above) would be resolved as well.

Of course, there needs to be acknowledgement that there is indeed a functional problem here, rather than this just being a matter of (1) incomplete documentation and (2) unjustified assumptions on the part of this coder.

**Comment by ashawley** [comment]
I'm not sure what can be done.  It seems we've narrowed down your issue.  Part of the problem is that combining multiple properties that are likely to succeed will result in a combined probability where they won't succeed.  You make a good case that there may be a way for ScalaCheck to interpret certain properties separately (or your words, "sequentially") rather than as a "compound" property to reduce the likelihood of a "surprising" outcome.

The _ScalaCheck_ book doesn't have anything specifically about discarded tests with `Prop.exists`, but there is a section in Chapter 4 on "Restricting test cases" which describes a related issue for implication ("pre-conditions") with `Prop.==>`.

> [...] Discarded test cases are not included in the total sum of passed tests for a property, which means that if the precondition is hard to fulfill when generating random test data, ScalaCheck might not be able to verify the property at all. Consider the following property:

```scala
import org.scalacheck.Prop.{forAll, BooleanOperators}
  val propSlice = forAll { (xs: List[Int], n: Int, m: Int) =>
      (m >= n && xs.indices.contains(n)
       && xs.indices.contains(m)
      ) ==> {
        val s = xs.slice(n,m)
        s.length == (m-n) && xs.containsSlice(s)
      }
}
```

```
scala> propSlice.check
  ! Gave up after only 2 passed tests. 99 tests were discarded.
```

> [...] As you can see, by randomly generating lists and integers ScalaCheck fails to come up with enough input sets that satisfy the precondition. In such cases, we need to help ScalaCheck a bit, and be more explicit about how the parameters should be generated. We can do this by nesting two calls to `forAll`, and use explicit generators to put boundaries on the generated integers:

```scala
  import org.scalacheck.Prop.{forAll, BooleanOperators}
  import org.scalacheck.Gen.oneOf
  val propSlice = forAll { xs: List[Int] =>
    forAll(oneOf(xs.indices), oneOf(xs.indices)) { (m, n) =>
      m >= n ==> {
        val s = xs.slice(n,m)
        s.length == (m-n) && xs.containsSlice(s)
      }
    }
  }
```

**Comment by jonaskoelker** [comment]
I think what @pford19 might be saying, and what I'm saying right now, is that the obvious mental model of `forAll` and `exists`, given their names, is that they quantify over something—in particular, they quantify over a set of (by default 100) samples drawn (independently?) from some population generated by a `Gen[T]`. If this mental model were correct, then `forAll(...) || forAll(...)` and `exists(...) && exists(...)` would behave as expected, perhaps using `Prop.Result.success` as a boolean algebra homomorphism into `({false, true}, &&, ||)`, and due to mumble-mumble-closure-properties it would behave in the obvious way for more complex combinations as well.

In the simple cases this mental model is correct (i.e. accurately predicts all observations). I think a reasonable approach is to either (a) make this mental model correct in general; or (b) rename `forAll` and `exists` to avoid misleading people; or (c) attach a giant-ass warning sign[1] about the surprising algebraic properties. ([1]: see also https://xkcd.com/37/)

Two of these approaches look difficult and backwards-incompatible, and if we're not constrained by that I think making the mental model correct is the best choice. However, dependencies from particular samples in outer quantifiers to choices of distributions in inner quantifiers might make this tricky. If I was green-fielding this I would probably not allow properties to nest (what does the nesting even mean?) and instead require the dependency to be expressed in the data generation, but I'm not :-)

By the way, what's the use case for `exists` and `||`? If you have a constructive existence proof of whatever, you could replace existential quantifiers with an implementation of the construction algorithm. Is `exists` primarily for non-constructive proofs? But doesn't it embody a constructive proof itself? [stochastic proof: case analysis over the entire range of the `Gen[T]`.]

If `||` succeeds in the context of `forAll(...) || forAll(...)` and each universal quantifier can fail, then `||` only succeeds by being lucky the following way: when the left quantifier fails the right quantifier succeeds and vice versa. But if the sampling is independent you have no way of guaranteeing this. So presumably `||` is either for coupled (non-independent) sampling, or for situations where one property always holds and the other sometimes fails. But in the latter case, couldn't you just drop the sometimes-failing property? Isn't the whole expression equivalent to the always-succeeding property?

And if the sampling is dependent, couldn't you rewrite the test into a `forAll(coupledSample) { leftProperty(extractLeftData(coupledSample)) || rightProperty(extractRightData(coupledSample)) }` and be back in the land of independent samples?

And having eliminated `exists` and `||` and dependent samples, couldn't you bubble up all your universal quantifiers, and then merge them into a single universal quantification over a tuple, and split uses of `Prop.&&` into multiple distinct properties?

So, maybe the problem is one of (a) using `exists`; or (b) using `||`; or (c) using a generator which does not generate samples independently, and the problem could be solved by the user if they apply the rewrites I suggested above.

I hope I offered food for thought.
